File: 547.number-of-provinces.java
import java.util.ArrayList;
import java.util.Arrays;

/*
 * @lc app=leetcode id=547 lang=java
 *
 * [547] Number of Provinces
 */

// @lc code=start
class Solution {

    boolean[] visited;



    public void DFS(int row, int[][] graph) {

        for (int col = 0; col < graph.length; col++) {
            if (graph[row][col] == 1 && !visited[col]) {

                visited[col] = true;
                DFS(col, graph);
            }
        }

    }

    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        if (n == 0 ||   isConnected[0].length == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        visited = new boolean[n];
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                DFS(i, isConnected);
                count++;
            }
        }
        return count;

    }
}
// @lc code=end

File: 547.number-of-provinces.py
#
# @lc app=leetcode id=547 lang=python3
#
# [547] Number of Provinces
#

# @lc code=start
from typing import List


class Solution:
    
    def dfs(self,current_node):
        self.visited[current_node] = True
            
        for nodes in range(self.length):
            if self.isConnected[current_node][nodes] and not self.visited[nodes]:
                self.dfs(nodes)
        
            
            
            
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        self.length = len(isConnected)
        self.isConnected = isConnected
        self.visited = [False] * self.length
        
        result = 0
        for i in range(self.length):
            if not self.visited[i]:
                self.dfs(i)
                result += 1
        return result
        
        
        
        
# @lc code=end

